### DISCUSSION TITLE
looking for example code in python to control hero 10 Black via usb

### DISCUSSION THREAD

**Discussion by AlanEbell** [gopro/OpenGoPro#146] [Q&A]
Hello,

I will gladly pay someone to write me an example in python to control a hero 10 on a Raspberry Pi 4 via USB.

The code simple needs to connect to the camera and start recording video for a predetermined time.  Then upload the video clip to the raspberry pi and delete it off the camera.

If anyone on here can do that for me I will gladly pay for your time.  let me know Thanks in advance.

Alan Bell

**Comment by KonradIT** [comment]
GoPro has not updated their Python package to support USB control, but I have updated my unofficial GoPro control library to support Hero9's gpWebcam USB API and Hero10 OpenGoPro USB API/gpWebcam stuff.

So by using [my python package](https://github.com/KonradIT/gopro-py-api), and taking [this example code](https://github.com/KonradIT/gopro-py-api/blob/master/examples/opengopro-record_video.py):

```python
import sys
import time
from goprocam import GoProCamera, constants
import threading
from signal import signal, SIGINT

record_len = 10 # How long to record for in seconds

def record_video(interface: str) -> None:

    gopro = GoProCamera.GoPro(ip_address=GoProCamera.GoPro.getWebcamIP(
        interface), camera=constants.gpcontrol, webcam_device=interface, api_type=constants.ApiServerType.OPENGOPRO)
    try:
        r = gopro.setWiredControl(constants.on)
        gopro.checkResponse(r)
    except:
        pass  # sometimes throws 500 server error when camera is already on wired control mode

    r = gopro.shoot_video(record_len)
    print(r)
    print("Video recorded")
    gopro.downloadLastMedia()
    gopro.delete(constants.Delete.LAST)
    exit()


def handler(s, f):
    thr.stop()
    quit()


signal(SIGINT, handler)

cameras = sys.argv[1]
cameras = cameras.split(",")

for interface in cameras:
    thr = threading.Thread(target=record_video, args=(interface,))
    thr.start()
```

**Reply by AlanEbell** [reply]
Thank you Konrad I appreciate it.

I think (because I am currently on windows 10) that I am having some difficulty getting the system to recognize my camera so it's not yet working for me, but I will work at it when I get my Raspberry Pi setup soon.

" cameras = sys.argv[1]
IndexError: list index out of range"

I assume this indicates that the script is not seeing the camera?

Warm Regards
Alan

**Reply by KonradIT** [reply]
Hi, you need to pass the interface name, use the script on my repo to get it: https://github.com/KonradIT/gopro-py-api/blob/master/examples/get_iface_name_win.py

Should look like an UUID. Pass it as an argument.

**Reply by KonradIT** [reply]
On the RPI it should look like /dev/xxxx.

**Reply by AlanEbell** [reply]
Thanks its weird but for what ever reason it's not seeing the camera.  I just did a factory reset so we will see what happens next.

Thanks

Alan

**Reply by KonradIT** [reply]
Weird, so the get_iface_name_win script does not output anything? Your GoPro must be on the USB screen (front screen black with USB logo)

```bash
❯ python .\examples\get_iface_name_win.py
IP: 172.20.122.52 from Interface {4567E1C0-2449-4706-80C6-839ACA1EA0A8}
```

**Reply by AlanEbell** [reply]
yes exactly.  I have tried everything maybe I will need to reinstall the firmware on the camera:  Notice in the device manager that the camera shows up as a device but the secondary connection is listed as CDC NCM and does not appear to be configured properly.

![hero10blackusbIssue](https://user-images.githubusercontent.com/13323211/157497747-ac797a98-3fec-488c-864e-919db52ed8f0.JPG)
?

**Reply by KonradIT** [reply]
Try installing the GoPro Webcam program, if not already installed: https://community.gopro.com/s/article/GoPro-Webcam?language=en_US

You should see a Network adapter called `UsbNcm Host Device`

**Reply by AlanEbell** [reply]
Can you tell me where I am going wrong with this code?  (the interface is now being found but I still get the arg out of range error
   cameras = sys.argv[1]
IndexError: list index out of range
python-BaseException


----------------------------------------------------------------------------------

import sys
import time
from goprocam import GoProCamera, constants
import threading
from signal import signal, SIGINT

record_len = 10 # How long to record for in seconds

def record_video(interface: str) -> None:

    gopro = GoProCamera.GoPro(ip_address=GoProCamera.GoPro.getWebcamIP(
        interface), camera=constants.gpcontrol, webcam_device=interface, api_type=constants.ApiServerType.OPENGOPRO)
    try:
        r = gopro.setWiredControl(constants.on)
        gopro.checkResponse(r)
    except:
        pass  # sometimes throws 500 server error when camera is already on wired control mode

    r = gopro.shoot_video(record_len)
    print(r)
    print("Video recorded")
    gopro.downloadLastMedia()
    gopro.delete(constants.Delete.LAST)
    exit()


def handler(s, f):
    thr.stop()
    quit()


signal(SIGINT, handler)

cameras = sys.argv[1]
cameras = cameras.split(",")



for interface in cameras:
    thr = threading.Thread(target=record_video, args=(interface,))
    thr.start()

if __name__ == '__main__':
    record_video('172.21.130.52')

**Reply by KonradIT** [reply]
You need to pass the UUID (in my previous example it's `{4567E1C0-2449-4706-80C6-839ACA1EA0A8}`) as a command line argument when calling the scritpt.

For instance: `python3 gopro-usb-record.py "{4567E1C0-2449-4706-80C6-839ACA1EA0A8}"`

The last two lines are not needed.

**Comment by AlanEbell** [comment]
but what if I don't plan to call it from the command line and need to do it on the module level and be able to run the program from within pycharm?

**Reply by KonradIT** [reply]
Much simpler then

```python
from goprocam import GoProCamera, constants

def record_video(interface: str) -> None:

    gopro = GoProCamera.GoPro(ip_address=GoProCamera.GoPro.getWebcamIP(
        interface), camera=constants.gpcontrol, webcam_device=interface, api_type=constants.ApiServerType.OPENGOPRO)
    try:
        r = gopro.setWiredControl(constants.on)
        gopro.checkResponse(r)
    except:
        pass  # sometimes throws 500 server error when camera is already on wired control mode

    r = gopro.shoot_video(10)
    print(r)
    print("Video recorded")
    gopro.downloadLastMedia()
    r = gopro.delete(constants.Delete.LAST)
    gopro.checkResponse(r)
    gopro.mode(constants.Mode.VideoMode)
    exit()

record_video("{4567E1C0-2449-4706-80C6-839ACA1EA0A8}")
```

**Reply by AlanEbell** [reply]
Thank you so much!  It works and now I have something to work off of.  I can not tell you how much I appreciate your help with this.  I tried to sponsor you again with another donation but for some reason since I already sponsored you once it say I cannot.

Wierd let me know if there is a way to unlock that and I will get back on and give you some more sponsorship thanks.

**Reply by KonradIT** [reply]
No idea how to sort that out, maybe try using paypal (button is on my website: https://konraditurbe.dev/hacking.html).

**Reply by AlanEbell** [reply]
This code on a raspberry pi 4 just sits and waits with a message that says "Waking up..."

from goprocam import GoProCamera, constants
import datetime

today = datetime.datetime.now()
date_time = today.strftime("%m_%d_%Y_%H:%M:%S")
def record_video(interface: str, duration) -> None:
    gopro = GoProCamera.GoPro(ip_address=GoProCamera.GoPro.getWebcamIP(interface), camera=constants.gpcontrol, webcam_device=interface, api_type=constants.ApiServerType.OPENGOPRO)
    try:
        r = gopro.setWiredControl(constants.on)
        gopro.checkResponse(r)
    except:
        pass  # sometimes throws 500 server error when camera is already on wired control mode

    r = gopro.shoot_video(duration)
    print(r)
    print("Video recorded")
    gopro.downloadLastMedia()
    r = gopro.delete(constants.Delete.LAST)
    gopro.checkResponse(r)
    gopro.mode(constants.Mode.VideoMode)
    exit()



if __name__ == '__main__':
    record_video("172.21.130.52",5)

**Reply by KonradIT** [reply]
For RPI, the `interface` variable will need to be a Linux kernel device identifier, such as `/dev/usb0` or similar. https://github.com/gopro/OpenGoPro/discussions/146#discussioncomment-2319331

Run `ls /dev/*` and find the USB Ethernet network. (I usually run the command, plug the GoPro, wait a few seconds, run the command again and check which device is present on the second run)

Then replace `172.21.130.52` with the device identifier.

**Comment by AlanEbell** [comment]
weird not seeing anything like that:

Did you get my email?  that I sent yesterday to the contact link of your website?

pi@raspberrypi:~ $ ls /dev/*
/dev/autofs         /dev/ram14           /dev/tty3       /dev/uinput
/dev/btrfs-control  /dev/ram15           /dev/tty30      /dev/urandom
/dev/cachefiles     /dev/ram2            /dev/tty31      /dev/vchiq
/dev/console        /dev/ram3            /dev/tty32      /dev/vcio
/dev/cuse           /dev/ram4            /dev/tty33      /dev/vc-mem
/dev/fb0            /dev/ram5            /dev/tty34      /dev/vcs
/dev/full           /dev/ram6            /dev/tty35      /dev/vcs1
/dev/fuse           /dev/ram7            /dev/tty36      /dev/vcs2
/dev/gpiochip0      /dev/ram8            /dev/tty37      /dev/vcs3
/dev/gpiochip1      /dev/ram9            /dev/tty38      /dev/vcs4
/dev/gpiomem        /dev/random          /dev/tty39      /dev/vcs5
/dev/hidraw0        /dev/rfkill          /dev/tty4       /dev/vcs6
/dev/hidraw1        /dev/rpivid-h264mem  /dev/tty40      /dev/vcs7
/dev/hidraw2        /dev/rpivid-hevcmem  /dev/tty41      /dev/vcsa
/dev/hwrng          /dev/rpivid-intcmem  /dev/tty42      /dev/vcsa1
/dev/initctl        /dev/rpivid-vp9mem   /dev/tty43      /dev/vcsa2
/dev/kmsg           /dev/serial1         /dev/tty44      /dev/vcsa3
/dev/libmtp-1-1.1   /dev/stderr          /dev/tty45      /dev/vcsa4
/dev/log            /dev/stdin           /dev/tty46      /dev/vcsa5
/dev/loop0          /dev/stdout          /dev/tty47      /dev/vcsa6
/dev/loop1          /dev/tty             /dev/tty48      /dev/vcsa7
/dev/loop2          /dev/tty0            /dev/tty49      /dev/vcsm-cma
/dev/loop3          /dev/tty1            /dev/tty5       /dev/vcsu
/dev/loop4          /dev/tty10           /dev/tty50      /dev/vcsu1
/dev/loop5          /dev/tty11           /dev/tty51      /dev/vcsu2
/dev/loop6          /dev/tty12           /dev/tty52      /dev/vcsu3
/dev/loop7          /dev/tty13           /dev/tty53      /dev/vcsu4
/dev/loop-control   /dev/tty14           /dev/tty54      /dev/vcsu5
/dev/media0         /dev/tty15           /dev/tty55      /dev/vcsu6
/dev/media1         /dev/tty16           /dev/tty56      /dev/vcsu7
/dev/mem            /dev/tty17           /dev/tty57      /dev/vga_arbiter
/dev/mmcblk0        /dev/tty18           /dev/tty58      /dev/vhci
/dev/mmcblk0p1      /dev/tty19           /dev/tty59      /dev/video10
/dev/mmcblk0p2      /dev/tty2            /dev/tty6       /dev/video11
/dev/null           /dev/tty20           /dev/tty60      /dev/video12
/dev/port           /dev/tty21           /dev/tty61      /dev/video13
/dev/ppp            /dev/tty22           /dev/tty62      /dev/video14
/dev/ptmx           /dev/tty23           /dev/tty63      /dev/video15
/dev/ram0           /dev/tty24           /dev/tty7       /dev/video16
/dev/ram1           /dev/tty25           /dev/tty8       /dev/video18
/dev/ram10          /dev/tty26           /dev/tty9       /dev/watchdog
/dev/ram11          /dev/tty27           /dev/ttyAMA0    /dev/watchdog0
/dev/ram12          /dev/tty28           /dev/ttyprintk  /dev/zero
/dev/ram13          /dev/tty29           /dev/uhid

/dev/block:
1:0  1:10  1:12  1:14  1:2  1:4  1:6  179:0  179:2  1:9  7:1  7:3  7:5  7:7
1:1  1:11  1:13  1:15  1:3  1:5  1:7  179:1  1:8    7:0  7:2  7:4  7:6

/dev/bus:
usb

/dev/char:
10:130  13:32    189:3    243:2  4:13  4:27  4:40  4:54  5:1    7:4   81:3
10:183  13:63    1:9      246:0  4:14  4:28  4:41  4:55  5:2    7:5   81:4
10:229  13:64    204:64   247:0  4:15  4:29  4:42  4:56  5:3    7:6   81:5
10:235  13:65    226:0    248:0  4:16  4:3   4:43  4:57  7:0    7:64  81:6
10:237  13:66    226:1    250:0  4:17  4:30  4:44  4:58  7:1    7:65  81:7
10:242  1:4      226:128  253:0  4:18  4:31  4:45  4:59  7:128  7:66
10:61   1:5      236:0    253:1  4:19  4:32  4:46  4:6   7:129  7:67
10:62   162:0    236:1    254:0  4:2   4:33  4:47  4:60  7:130  7:68
10:63   1:7      237:0    254:1  4:20  4:34  4:48  4:61  7:131  7:69
1:1     1:8      238:0    29:0   4:21  4:35  4:49  4:62  7:132  7:7
1:11    189:0    239:0    4:0    4:22  4:36  4:5   4:63  7:133  7:70
116:0   189:1    240:0    4:1    4:23  4:37  4:50  4:7   7:134  7:71
116:16  189:12   242:0    4:10   4:24  4:38  4:51  4:8   7:135  81:0
116:33  189:128  243:0    4:11   4:25  4:39  4:52  4:9   7:2    81:1
1:3     189:2    243:1    4:12   4:26  4:4   4:53  5:0   7:3    81:2

/dev/disk:
by-id  by-label  by-partuuid  by-path  by-uuid

/dev/dma_heap:
linux,cma  system

/dev/dri:
by-path  card0  card1  renderD128

/dev/fd:
0  1  2  3

/dev/input:
by-id  by-path  event0  event1  event2  mice  mouse0

/dev/mapper:
control

/dev/mqueue:

/dev/net:
tun

/dev/pts:
0  ptmx

/dev/raw:
rawctl

/dev/shm:

/dev/snd:
by-path  controlC0  pcmC0D0p  seq  timer

/dev/v4l:
by-path

**Reply by AlanEbell** [reply]
its strange I am finding this line which looks like it could be it but I still with "Waking up...."  
dev/libmtp-1-1.1 -> bus/usb/001/015

**Reply by shahla-ai** [reply]
Hello @AlanEbell 
did you find a solution to run on RPI ?
